python/lsst/daf/butler/registry/interfaces/_datasets.py

```diff
@@ -701,3 +381,292 @@ def fetch_summaries(
+            With non-`UNIQUE` mode ID is computed from some combination of
+            dataset type, dataId, and run collection name; if the same ID is
+            already in the database then new record is not inserted.
+
+        Returns
+        -------
+        datasets : `list` [ `DatasetRef` ]
+            References to the inserted datasets.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def import_(
+        self,
+        dataset_type_name: str,
+        run: RunRecord,
+        data_ids: Mapping[DatasetId, DataCoordinate],
+    ) -> list[DatasetRef]:
+        """Insert one or more dataset entries into the database.
+
+        Parameters
+        ----------
+        dataset_type_name : `str`
+            Name of the dataset type.
+        run : `RunRecord`
+            The record object describing the `~CollectionType.RUN` collection
+            these datasets will be associated with.
+        data_ids : `~collections.abc.Mapping`
+            Mapping from dataset ID to data ID.
+
+        Returns
+        -------
+        datasets : `list` [ `DatasetRef` ]
+            References to the inserted or existing datasets.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def delete(self, datasets: Iterable[DatasetId | DatasetRef]) -> None:
+        """Fully delete the given datasets from the registry.
+
+        Parameters
+        ----------
+        datasets : `~collections.abc.Iterable` [ `DatasetId` or `DatasetRef` ]
+            Datasets to be deleted.  If `DatasetRef` instances are passed,
+            only the `DatasetRef.id` attribute is used.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def associate(
+        self, dataset_type: DatasetType, collection: CollectionRecord, datasets: Iterable[DatasetRef]
+    ) -> None:
+        """Associate one or more datasets with a collection.
+
+        Parameters
+        ----------
+        dataset_type : `DatasetType`
+            Type of all datasets.
+        collection : `CollectionRecord`
+            The record object describing the collection.  ``collection.type``
+            must be `~CollectionType.TAGGED`.
+        datasets : `~collections.abc.Iterable` [ `DatasetRef` ]
+            Datasets to be associated.  All datasets must have the same
+            `DatasetType` as ``dataset_type``, but this is not checked.
+
+        Notes
+        -----
+        Associating a dataset with into collection that already contains a
+        different dataset with the same `DatasetType` and data ID will remove
+        the existing dataset from that collection.
+
+        Associating the same dataset into a collection multiple times is a
+        no-op, but is still not permitted on read-only databases.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def disassociate(
+        self, dataset_type: DatasetType, collection: CollectionRecord, datasets: Iterable[DatasetRef]
```

Comment:
This is one of the cases I just didn't want to add to the scope of the ticket, because while you're right, there was just so much of it that it would have been a distraction.